### PR TITLE
Introduce Mapping-like protocols

### PR DESCRIPTION
**Pull request by srittau**
typing.Mapping is not a protocol, which has caused problems in the past.
(E.g. python/typeshed#3569, see also python/typeshed#3576.) This
introduces a few narrow protocols to _typeshed.pyi that can be used in
place of Mapping.

Not all uses of Mapping can be replaced. For example, cgi.FieldStorage
explictly checks whether the supplied headers argument is a Mapping
instance.

### CODE REVIEW

File: stdlib/2and3/_typeshed/__init__.pyi [hunk 1]
```diff
@@ -15,12 +15,28 @@
...
 _T_co = TypeVar("_T_co", covariant=True)
 _T_contra = TypeVar("_T_contra", contravariant=True)
 
+# Mapping-like protocols
+
+class SupportsItems(Protocol[_KT_co, _VT_co]):
+    def items(self) -> AbstractSet[Tuple[_KT_co, _VT_co]]: ...
+
+class SupportsGetItem(Container[_KT_contra], Protocol[_KT_contra, _VT_co]):
+    def __getitem__(self, __k: _KT_contra) -> _VT_co: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/_typeshed/__init__.pyi:34]: I feel like we had bad experiences in the past making Mapping anything other than invariant in the key type, but can't find those discussions now

Comment (JelleZijlstra) [inline stdlib/2and3/_typeshed/__init__.pyi:34]: See python/typing#273 and #550 for some previous discussion on the variance of Mapping (not directly relevant here).

**Comment by Akuli** [comment]
For example, `TopologicalSorter.__init__` takes a mapping and only uses the `.items()` method. What happens if it's changed to instead loop over the mapping to get the keys and call `__getitem__` to get the values in a future version of Python? Then requiring `SupportsItems` instead of `Mapping` will be incorrect.

**Comment by srittau** [comment]
If the implementation changes, the stubs need to change as well. We tend to follow implementation, not documentation in typeshed.

**Comment by torstenrudolf** [comment]
is there a way I can use the `SupportsGetItem` in type annotations in my code? I cannot figure out how to import it.

**Comment by hauntsaninja** [comment]
Look at the comment at the top of the file :-)

**Pull request merged**